pyiceberg/partitioning.py

```diff
@@ -272,6 +272,60 @@ def assign_fresh_partition_spec_ids(spec: PartitionSpec, old_schema: Schema, fre
 T = TypeVar("T")
 
 
+class PartitionMap(Generic[T]):
+    _specs: dict[int, PartitionSpec]
+    _partition_maps: dict[int, dict[Record | None, T]]
+
+    def __init__(self, specs: dict[int, PartitionSpec]):
+        self._specs = specs
+        self._partition_maps = {}
+
+    def __len__(self) -> int:
+        """Return the length of the partition map.
+
+        Returns:
+            length of _partition_maps
+        """
+        return len(self.values())
+
+    def is_empty(self) -> bool:
+        return len(self._partition_maps.values()) == 0
+
+    def contains_key(self, spec_id: int, struct: Record) -> bool:
+        return self._partition_maps.get(spec_id) is not None
+
+    def contains_value(self, value: T) -> bool:
+        return value in self._partition_maps.values()
```

Comment:
If we were to follow the [Java reference implementation](https://github.com/apache/iceberg/blob/026374b530eeea46429cef61a99ba9321e0587b6/core/src/main/java/org/apache/iceberg/util/PartitionMap.java#L89), we'd want to check if the `value` exists in all the contained partition map values. (`self._partition_maps` is a dict of a dict)

```suggestion
        return any(value in partition_map.values() for partition_map in self._partition_maps.values())
```